### PR TITLE
can schedule endpoint

### PR DESCRIPTION
**Pull request by chkeita**
## Summary of the Pull Request

_What is this about?_

## PR Checklist
* [ ] Applies to work item: #xxx
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/onefuzz) and sign the CLI.
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_

## Validation Steps Performed

_How does someone test & validate?_

### CODE REVIEW

File: src/api-service/__app__/agent_can_schedule/__init__.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    node = Node.get_by_machine_id(request.machine_id)
+    if not node:
+        return not_ok(
+            Error(code=ErrorCode.UNABLE_TO_FIND, errors=["unable to find node"]),
+            context=request.machine_id,
+        )
+
+    allowed = True
+    work_stopped = False
+    if node.version != current_version:
```
Comment (bmc-msft) [inline src/api-service/__app__/agent_can_schedule/__init__.py:34]: Can this be a function of Node?  I imagine we'll likely want to do this in multiple places.

File: src/api-service/__app__/agent_can_schedule/__init__.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+from ..onefuzzlib.request import not_ok, ok, parse_uri
+from ..onefuzzlib.tasks.main import Task
+
+
+def get(req: func.HttpRequest) -> func.HttpResponse:
+    request = parse_uri(CanScheduleRequest, req)
+    if isinstance(request, Error):
+        return not_ok(request, context="CanScheduleRequest")
+
+    current_version = ""
```
Comment (bmc-msft) [inline src/api-service/__app__/agent_can_schedule/__init__.py:23]: onefuzzlib.__version__.__version__

File: src/api-service/__app__/agent_can_schedule/__init__.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    if not node:
+        return not_ok(
+            Error(code=ErrorCode.UNABLE_TO_FIND, errors=["unable to find node"]),
+            context=request.machine_id,
+        )
+
+    allowed = True
+    work_stopped = False
+    if node.version != current_version:
+        allowed = False
```
Comment (ranweiler) [inline src/api-service/__app__/agent_can_schedule/__init__.py:35]: It'd probably be nice to have some logging here

File: src/api-service/__app__/agent_can_schedule/__init__.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        stop_message.save()
+
+    task = Task.get_by_task_id(request.task_id)
+
+    work_stopped = isinstance(task, Error) or (task.state == TaskState.scheduled)
+    return ok(CanSchedule(allowed=allowed, work_stopped=work_stopped))
+
+
+def main(req: func.HttpRequest) -> func.HttpResponse:
+    if req.method == "GET":
```
Comment (ranweiler) [inline src/api-service/__app__/agent_can_schedule/__init__.py:49]: We should probably make this a `POST`. `GET` is really meant for requests with no body. I'll update my PR too.

File: src/api-service/__app__/agent_can_schedule/function.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
+{
+	"scriptFile": "__init__.py",
+	"bindings": [
+		{
+			"authLevel": "anonymous",
+			"type": "httpTrigger",
+			"direction": "in",
+			"name": "req",
+			"methods": [
```
Comment (ranweiler) [inline src/api-service/__app__/agent_can_schedule/function.json:9]: For now, I suggest just not routing the unused methods (I assume these are just here from copy/paste).

File: src/api-service/__app__/agent_can_schedule/__init__.py (hunk 4 above)
Comment (ranweiler) [reply src/api-service/__app__/agent_can_schedule/__init__.py:49]: (done)

File: src/api-service/__app__/agent_can_schedule/__init__.py (hunk 1 above)
Comment (chkeita) [reply src/api-service/__app__/agent_can_schedule/__init__.py]: moves logic to Node class

File: src/api-service/__app__/agent_can_schedule/__init__.py (hunk 2 above)
Comment (chkeita) [reply src/api-service/__app__/agent_can_schedule/__init__.py]: moves logic to Node class

File: src/api-service/__app__/agent_can_schedule/__init__.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        )
+        allowed = False
+        stop_message = NodeMessage(
+            agent_id=node.machine_id, message=NodeCommand(stop=StopNodeCommand()),
+        )
+        stop_message.save()
+
+    task = Task.get_by_task_id(request.task_id)
+
+    work_stopped = isinstance(task, Error) or (task.state == TaskState.scheduled)
```
Comment (ranweiler) [inline src/api-service/__app__/agent_can_schedule/__init__.py:47]: Just double-checking: if the task isn't found, that is represented as an `Error`, not a `None`?

Also, I think this should be `task.state != TaskState.scheduled`.

Comment (chkeita) [reply src/api-service/__app__/agent_can_schedule/__init__.py]: yup here is the def get_by_task_id currently returns an error when the task is not found

**Review by ranweiler — APPROVED** [review]

**Pull request merged**